src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+register_dsp_command([0x98], 0);
+
+// 8-bit high-speed single-cycle DMA mode digitized sound input.
+register_dsp_command([0x99], 0);
+
+// Set input mode to mono.
+register_dsp_command([0xA0], 0);
+
+// Set input mode to stereo.
+register_dsp_command([0xA8], 0);
+
+// Program 16-bit DMA mode digitized sound I/O.
+register_dsp_command(any_first_digit(0xB0), 3, function()
+{
+    if(this.command & (1 << 3))
+    {
+        // Analogue to digital not implemented.
+        this.dsp_default_handler();
+        return;
+    }
+    var mode = this.write_buffer.shift();
+    this.dma_irq = SB_IRQ_16BIT;
+    this.dma_channel = this.dma_channel_16bit;
+    this.dma_autoinit = !!(this.command & (1 << 2));
+    this.dsp_signed = !!(mode & (1 << 4));
+    this.dsp_stereo = !!(mode & (1 << 5));
+    this.dsp_16bit = true;
+    this.dma_transfer_size_set();
+    this.dma_transfer_start();
+});
+
+// Program 8-bit DMA mode digitized sound I/O.
+register_dsp_command(any_first_digit(0xC0), 3, function()
+{
+    if(this.command & (1 << 3))
+    {
+        // Analogue to digital not implemented.
+        this.dsp_default_handler();
+        return;
+    }
+    var mode = this.write_buffer.shift();
+    this.dma_irq = SB_IRQ_8BIT;
+    this.dma_channel = this.dma_channel_8bit;
+    this.dma_autoinit = !!(this.command & (1 << 2));
+    this.dsp_signed = !!(mode & (1 << 4));
+    this.dsp_stereo = !!(mode & (1 << 5));
+    this.dsp_16bit = false;
+    this.dma_transfer_size_set();
+    this.dma_transfer_start();
+});
+
+// Pause 8-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD0], 0);
+
+// Turn on speaker.
+// Documented to have no effect on SB16.
+register_dsp_command([0xD1], 0, function()
+{
+    this.dummy_speaker_enabled = true;
+});
+
+// Turn off speaker.
+// Documented to have no effect on SB16.
+register_dsp_command([0xD3], 0, function()
+{
+    this.dummy_speaker_enabled = false;
+});
+
+// Continue 8-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD4], 0);
+
+// Pause 16-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD5], 0);
+
+// Continue 16-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD6], 0);
+
+// Get speaker status.
+register_dsp_command([0xD8], 0, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(this.dummy_speaker_enabled * 0xFF);
+});
+
+// Exit 16-bit auto-init DMA mode digitized sound I/O.
+// Exit 8-bit auto-init mode digitized sound I/O.
+register_dsp_command([0xD9, 0xDA], 0, function()
+{
+    this.dma_autoinit = false;
+});
+
+// DSP identification
+register_dsp_command([0xE0], 1, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(~this.write_buffer.shift());
+});
+
+// Get DSP version number.
+register_dsp_command([0xE1], 0, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(4);
+    this.read_buffer.push(5);
+});
+
+// DMA identification - based completely off dosbox
```

Comment:
We can't copy or port code from dosbox, because their license (GPL) isn't compatible with ours (BSD2). Please delete this code and write it again from scratch.

Reply:
Yikes. I'll be careful with licenses in the future. My implementation of this DMA identification routine is already a bit different to the one dosbox's uses, but I'll remove this routine entirely because:
1. I haven't encountered a situation that uses this routine yet.
2. I can't find any documentation about this routine.
3. I don't really understand this routine myself so I can't rewrite it from scratch anyway :stuck_out_tongue: